### ISSUE TITLE
Some questions related to generalisation

### ISSUE THREAD

**Issue by nad** [agda/agda#3117]
* Should it be possible to give "generalised arguments" by position? (No?)
* Should it be possible to give "hierarchical generalised arguments" by name? (Yes?)
* The algorithm currently chooses one hierarchical name, even though other names might also make sense. Should all of them be allowed when giving arguments by name? (Yes?)

**Comment by nad** [comment]
* Should it be possible to export generalisable names?

  In that case it should be possible to mark the names as being private. As this is currently not possible I think we should make a decision before the next release, to avoid changing the meaning of code in a later release.

**Comment by UlfNorell** [comment]
> Should it be possible to export generalisable names?

This is one of the todo's from the PR and should have it's own issue.

**Comment by WolframKahl** [comment]
* Is `generalize` really the right name for this?

  A similar feature has been part of many algebraic specification formalisms for ages; there the keyword usually is `vars` respectively `var` (CASL, for example, allows both, independent of the number of items).

  My proposal would be `variables` and `variable`.

**Comment by UlfNorell** [comment]
I wouldn't mind `variable` (singular; we don't have plural versions of `postulate` and `field`).

**Comment by nad** [comment]
Coq uses the keywords `Variable`, `Variables`, `Hypothesis` and `Hypotheses` in sections (see the [Coq manual](https://coq.inria.fr/refman/language/gallina-specification-language.html#assumptions)). As discussed elsewhere Coq sections are somewhat different from the Agda feature discussed here. However, I think the two concepts are sufficiently similar that it makes sense to use an established term  like "variable" as the keyword.

**Comment by andreasabel** [comment]
> Coq uses the keywords Variable, Variables, Hypothesis and Hypotheses in sections 

These are our module parameters.
The generalization feature is what is called _Implicit Generalization_ (search for this header here: https://coq.inria.fr/refman/language/gallina-extensions.html).

I think reusing `variable` or `hypothesis` for the generalization feature would be confusing.  Note that
```
  generalize n : Nat
```
does not mean that the development is parametrized by a `n : Nat` or that you have such a variable in scope.

**Comment by WolframKahl** [comment]
> Note that
> ```
>   generalize n : Nat
> ```
> does not mean that the development is parametrized by a `n : Nat` or that you have such a variable in scope.

This is the same as how I would explain
```
   var n : Nat
```
in CASL: The scopes of `n` are where it is implicitly universally quantified.

**Comment by andreasabel** [comment]
In PVS `x : VAR A` gives the type of `x` should it be bound somewhere, and it seems to automatically forall-generalize in propositions, see: http://pvs.csl.sri.com/doc/pvs-language-reference.pdf#section.3.2 and http://pvs.csl.sri.com/doc/pvs-language-reference.pdf#chapter.6 .

Isabelle does automatic forall-generalization like ML without any pragma for the variables needed.

**Comment by Lapin0t** [comment]
Nice feature! Some comments/feature suggestions after my fiddling with it (not sure if it deserves an issue of it's own):
* variables get exported which may be intended but it's not very intuitive: you get generalization over random identifiers just by `open`ing some module (and thus possible name clashes); for me it should behave as an implementation details just to make the code cleaner and additionally it breaks with "only declarations get exported"
* `private` should be supported if variables can be exported
* don't we want "cascading" variables to be generalized over only once (left-to-right, in-to-out)? eg:
  ```
  variable
    {i} : Level
    {X} : Set i

  postulate
    f : X -> Set i 
    -- actual type: {i0 i1} {X : Set i0} -> X -> Set i1
    -- i expected:  {i} {X : Set i} -> X -> Set i
  ```

For the last point, it's a big change since if eg `variable {X Y} : Set i`, in my proposed change any type mentioning both `X` and `Y` will have them living at the same level, so the old behavior will be a bit more verbose with:
```
variable
  {i0 i1} : Level
  {X} : Set i0
  {Y} : Set i1
```
Still i think i would like it better since it is much more explicit and it actually allows using `variable` in more deep "cascading" things: currently, you can only `variable`ify up to the most inner thing *appearing only once*.

**Comment by Lapin0t** [comment]
By the way, some feature which might be interesting to have is some kind of pattern synonyms for generalized variables, eg you declare `i0` and `incr` as level variables and then declare pattern `i1 := i0 ⊔ incr `. This would very much reduce the pain of doing level generalization on structures that have lots of set arguments for which certain level inequalities must hold. Of course another way to solve it would be to have a special syntax to lay out some level variables topologies.

I'm not sure i have good ideas for a syntax but maybe something like normal declarations (`$id : $type \n $id = $expr`) or weird let clauses (`let $id = $expr`) inside the `variable` block.

Actually now that i think about it a bit more, this is probably just gonna get the `generalize` feature really messy (and i completely +1 on andreas' point in an early generalization feature request that this has a big potential to be a pain to maintain if it's hacked in). So maybe one way to really go forward is to work out some tools to work with expressions with (truly) free variables (eg typed macros), quasi-quoting etc. Note that similar things might already be doable with reflection (albeit with no syntactic sugar), i'm gonna explore that.

Sorry if this whole thing is out of topic btw, i'm gonna continue on the ML or in another thread if it's better.

**Comment by divipp** [comment]
> By the way, some feature which might be interesting to have is some kind of pattern synonyms for generalized variables, eg you declare i0 and incr as level variables and then declare pattern i1 := i0 ⊔ incr. This would very much reduce the pain of doing level generalization on structures that have lots of set arguments for which certain level inequalities must hold. Of course another way to solve it would be to have a special syntax to lay out some level variables topologies.

This has a separate issue now:  #3275

**Comment by divipp** [comment]
> private should be supported if variables can be exported

Implemented on the `generalize` branch.

**Comment by divipp** [comment]
> Some comments/feature suggestions after my fiddling with it (not sure if it deserves an issue of it's own)

See also #3163

**Comment by divipp** [comment]
Closed in favor of #3280
